sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsublite/SubscriptionPartitionLoader.java

```diff
@@ -0,0 +1,102 @@
+import com.google.cloud.pubsublite.PartitionLookupUtils;
+import com.google.cloud.pubsublite.SubscriptionPath;
+import com.google.cloud.pubsublite.TopicPath;
+import com.google.common.annotations.VisibleForTesting;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import org.apache.beam.sdk.transforms.Create;
+import org.apache.beam.sdk.transforms.MapElements;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.apache.beam.sdk.transforms.Watch;
+import org.apache.beam.sdk.transforms.Watch.Growth.PollFn;
+import org.apache.beam.sdk.transforms.Watch.Growth.PollResult;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.TypeDescriptor;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableList;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+
+class SubscriptionPartitionLoader extends PTransform<PBegin, PCollection<SubscriptionPartition>> {
+  private final TopicPath topic;
+  private final SubscriptionPath subscription;
+  private final SerializableFunction<TopicPath, Integer> getPartitionCount;
+  private final Duration pollDuration;
+  private final boolean terminate;
+
+  SubscriptionPartitionLoader(TopicPath topic, SubscriptionPath subscription) {
+    this(
+        topic,
+        subscription,
+        PartitionLookupUtils::numPartitions,
+        Duration.standardMinutes(1),
+        false);
+  }
+
+  @VisibleForTesting
+  SubscriptionPartitionLoader(
+      TopicPath topic,
+      SubscriptionPath subscription,
+      SerializableFunction<TopicPath, Integer> getPartitionCount,
+      Duration pollDuration,
+      boolean terminate) {
+    this.topic = topic;
+    this.subscription = subscription;
+    this.getPartitionCount = getPartitionCount;
+    this.pollDuration = pollDuration;
+    this.terminate = terminate;
+  }
+
+  @Override
+  public PCollection<SubscriptionPartition> expand(PBegin input) {
+    PCollection<TopicPath> start = input.apply(Create.of(ImmutableList.of(topic)));
+    PCollection<KV<TopicPath, Partition>> partitions =
+        start.apply(
+            Watch.growthOf(
+                    new PollFn<TopicPath, Partition>() {
+                      @Override
+                      public PollResult<Partition> apply(TopicPath element, Context c) {
+                        checkArgument(element.equals(topic));
+                        int partitionCount = getPartitionCount.apply(element);
+                        List<Partition> partitions =
+                            IntStream.range(0, partitionCount)
+                                .mapToObj(Partition::of)
+                                .collect(Collectors.toList());
+                        return PollResult.incomplete(Instant.EPOCH, partitions);
```

Comment:
Is it intended to always have the watermark as `Instant.EPOCH`? It will hold back the system watermark unnecessary and the worst case is that downstream operations will have to wait for this watch transform to complete to process.

Reply:
`Instant.now()` is better than `Instant.EPOCH`. You should care about the watermark here since the watermark here will affect the watermark of the whole pipeline.

Reply:
It shouldn't, since the next stage will overwrite the timestamps and watermarking IIUC.

Reply:
Done.

Reply:
>It shouldn't, since the next stage will overwrite the timestamps and watermarking IIUC.

At least Dataflow will compute a global watermark, which is the minimal watermark of all stages. That's why you don't want to hold the watermark here.